lib/RefBackend/RefBackend.cpp

```diff
@@ -95,28 +95,45 @@ static LogicalResult mungeFunction(
   SmallVector<Operation *> toErase;
   bool hadError = false;
   func.walk([&](ReturnOp op) {
-    auto memRefType = op.getOperandTypes()[0].dyn_cast<MemRefType>();
-    if (!memRefType) {
+    auto returnType = op.getOperandTypes()[0];
+    if (returnType.isa<MemRefType>()) {
+      auto elemType = returnType.dyn_cast<MemRefType>().getElementType();
+      auto unRankedMemrefType = UnrankedMemRefType::get(elemType, 0);
+      auto it = consumeFuncReturnFuncs.find(unRankedMemrefType);
+      if (op.getNumOperands() != 1 || it == consumeFuncReturnFuncs.end()) {
```

Comment:
nit: the `consumeFuncReturnFuncs.find` call and the error check is the same for every `if` branch. The only think that changes is what you pass to `find`, so this can probably be refactored.